File: lib/interop/stream_reader.rb
require 'interop/message'
require 'interop/reader'

module Hx
  module Interop
    # Reads messages from a stream (e.g. STDIN)
    class StreamReader
      include Reader

      # Acceptable line terminators
      NEWLINES = Set.new(%W[\n \r\n]).freeze

      # @param [IO, StringIO] stream
      def initialize(stream)
        @stream = stream
        @mutex  = Mutex.new
      end

      protected

      def _read
        @mutex.synchronize do
          message      = Message.new(read_headers)
          length       = message.headers[Headers::CONTENT_LENGTH]
          message.body = length.nil? || length.empty? ? read_paragraph.join : read_length(length.to_i)
          message
        end
      end

      private

      # @return [Hash]
      def read_headers
        read_paragraph.to_h do |line|
          line.strip.split(/:\s*/, 2).tap do |pair|
            raise Error::InvalidHeader unless pair.length == 2
          end
        end
      end

      def read_paragraph
        lines = []
        while (line = @stream.readline("\n"))
          return lines if NEWLINES.include? line

          lines << line
        end
      end

      def read_length(length)
        result = length.positive? ? @stream.read(length) : ''
        sep    = @stream.read(1)
        sep    = @stream.read(1) if sep == "\r"
        raise Error::Unexpected, "Expected a newline after #{length} bytes" unless sep == "\n"

        result
      end
    end
  end
end

File: php/Interop/StreamReader.php
<?php

namespace Hx\Interop;

use Hx\Interop\Error\Unexpected;

class StreamReader implements Reader {
    private ByteReader $byteReader;

    /**
     * @param ByteReader $byteReader
     */
    public function __construct(ByteReader $byteReader) {
        $this->byteReader = $byteReader;
    }

    public function read(): Message {
        $message = new Message($this->readHeaders());
        $length = $message[Header::CONTENT_LENGTH];
        $message->body = ($length === null || $length === '') ?
            join('', $this->readParagraph()) :
            $this->readLength(intval($length));
        return $message;
    }

    /**
     * @return Headers
     * @throws EOF
     * @throws Error\InvalidHeader
     */
    private function readHeaders(): Headers {
        return Headers::parse($this->readParagraph());
    }

    /**
     * @return string[]
     * @throws EOF
     */
    private function readParagraph(): array {
        $result = [];
        while (true) {
            $line = $this->readLine();
            if ($line === "\n" || $line === "\r\n") {
                return $result;
            }
            $result[]= $line;
        }
    }

    /**
     * @param int $length
     * @return string
     * @throws EOF
     * @throws Unexpected if the given length of bytes is not followed by a newline.
     */
    private function readLength(int $length): string {
        $result = $length >= 0 ? $this->byteReader->readBytes($length) : '';
        $sep = $this->byteReader->readBytes(1);
        if ($sep == "\r") {
            $sep = $this->byteReader->readBytes(1);
        }
        if ($sep !== "\n") {
            throw new Unexpected("Expected a newline after $length bytes");
        }
        return $result;
    }

    /**
     * @return string
     * @throws EOF
     */
    private function readLine(): string {
        return $this->byteReader->readUntil("\n");
    }
}
